allow constants to refer to mutable/external memory, but reject such constants as patterns

RalfJung opened the pull request:
Reference update for:

- https://github.com/rust-lang/rust/pull/140942

RalfJung committed:
allow constants to refer to mutable/external memory, but reject such constants as patterns

RalfJung commented:
So, I can't tell you where any of that is in the reference, I had a hard time even find the things I am editing here.^^

But I can tell you what the compiler does:

```rust
static mut S: i32 = 1;
const C: &mut i32 = unsafe { &mut S }; // ERROR: constructing invalid value
```
We do a type-directed pass over the final value of the constant, mostly to check e.g. that every `bool` we find is 0 or 1 and things like that. As part of that pass, we just throw an error if we find any mutable reference in a `const`. We also error if we find an `UnsafeCell` in memory that is immutable (see tests `tests/ui/statics/mutable_memory_validation.rs` and `tests/ui/consts/interior-mut-const-via-union.rs`) in either `const` or `static`, and if we find an `&mut` pointing to immutable memory (that last case might be just UB due to an aliasing violation, or it might be const being extra paranoid, that depends on the details of the aliasing model).

```rust
const C: &AtomicU32 = &AtomicU32::new(1);  // ERROR: constants cannot refer to interior mutable data
```
That falls out of the following two things which are in the reference:
- "shared borrows to values with interior mutability are only allowed to refer to *transient* places"
- and lifetime extension, which says that the `AtomicU32` constructed here is *not* transient

> Similarly, I'm trying to understand [E0764](https://doc.rust-lang.org/error_codes/E0764.html). Is the description of E0764 out of date?

Oh yeah, that is very outdated. (That's probably true for many of our error description. Multiple of our Exxxx tests don't even emit the right error code any more...)

```rust
const C: &mut i32 = &mut 1; // ERROR: mutable references are not allowed in the final value of constants
```
This is the exact same as "constants cannot refer to interior mutable data", but for a mutable reference rather than an interior mutable shared ref. I don't know why the error messages are worded so differently...

ehuss commented on src/items/constant-items.md:
````diff
@@ -48,9 +48,6 @@ const BITS_N_STRINGS: BitsNStrings<'static> = BitsNStrings {
 };
 ```
 
-r[items.const.final-value-immutable]
-The final value of a `const` item cannot contain references to anything mutable.
-
````
We were wondering if it would make sense to change the wording here instead of removing it. Perhaps something like:

"The final value of a `const` item cannot contain a mutable reference."

Most of us seemed to think this is correct. Perhaps there is some other rule forbidding this already (implicitly or explicitly), but it seems like it wouldn't hurt to be clear about it here?

traviscross committed:
Add clarifications

In the text as revised, it wasn't clear what rule prevented a mutable
reference from appearing in the final value of a constant item, so
let's add a rule specifically for that.

Similarly, it may not be immediately clear what rule prevents a
reference to a lifetime-extended interior mutable temporary created in
the initializer, so let's add an admonition section describing how
other rules apply to prevent this.

traviscross committed:
Revise editorially

RalfJung commented on src/items/constant-items.md:
```diff
@@ -49,7 +49,26 @@ const BITS_N_STRINGS: BitsNStrings<'static> = BitsNStrings {
 
 r[items.const.final-value-immutable]
-The final value of a `const` item cannot contain references to anything mutable.
+The final value of a `const` item cannot contain a mutable reference.
```
This is pretty much an entirely new clause, at least in terms of intent, so IMO it'd make sense to give it a new name, such as `items.const.no-mutable-ref`.

RalfJung commented on src/patterns.md:
```diff
@@ -975,6 +975,9 @@ r[patterns.const.generic]
 This means that associated consts that involve generic parameters cannot be used as patterns.
 
+r[patterns.const.immutable]
+The value of `C` must not contain any references to mutable statics (`static mut` or interior mutable) or `extern` statics.
```
```suggestion
The value of `C` must not contain any references to mutable statics (`static mut` or interior mutable `static`) or `extern` statics.
```
